### PR TITLE
Initial Commit For api-04-remote-and-local Objective

### LINKED ISSUE #42: api-04-remote-and-local / fetch data from our API and the process it locally
- [x] objective.json
- [x] description.md
- [x] validator.js
- [x] walkthrough.md

For more details, reference the [original objective planning doc](https://docs.google.com/document/d/1Esz69gd10vK4Btnk_pn0pmKTUgIq9f7sCrmnMVyOt5E/edit#heading=h.uoyv4k1cq39u).

## Objective outline
- Get all author names who have written a document longer than X pages where X is an argument passed into the function
- depends on #53

### PR DESCRIPTION
**Pull request by MarvelProgramming**
For issue #42

### CODE REVIEW

**Review by rmkubik — APPROVED** [review]

File: levels/lovelace_tower/objectives/api-04-remote-and-local/walkthrough.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
+# What is an API?
+
+[APIs](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Client-side_web_APIs/Introduction#what_are_apis) (Application Programming Interface) provide ways in which we can interact with other computer programs. They do so by publicly exposing what are called `endpoints`. Each `endpoint` represents a different service, and made accessible via a unique URL (Uniform Resource Locator) and [method](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods).
```
Comment (joenash) [inline levels/lovelace_tower/objectives/api-04-remote-and-local/walkthrough.md]:
```suggestion
[APIs](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Client-side_web_APIs/Introduction#what_are_apis) (Application Programming Interface) provide a way to interact with other computer programs. They do so by publicly exposing `endpoints`, with each `endpoint` representing a different service, and made accessible via a unique URL (Uniform Resource Locator) and [method](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods).
```

File: levels/lovelace_tower/objectives/api-04-remote-and-local/walkthrough.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+## What is JSON?
+
+[JSON](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON) (Javascript Object Notation) is a common way of formatting data, and is often used when making requests to endpoints that provide information in their response. It's important to note that when using `fetch` and receiving JSON, we must [parse](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON#static_methods) (i.e. have the computer anylize) the JSON data before using it.
```
Comment (joenash) [inline levels/lovelace_tower/objectives/api-04-remote-and-local/walkthrough.md]:
```suggestion
[JSON](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON) (Javascript Object Notation) is a common format used to represent and exchange data, and is often used when making requests to endpoints that provide information in their response. It's important to note that when using `fetch` and receiving JSON, we must [parse](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON#static_methods) (i.e. have the computer anylize) the JSON data before using it.
```

**Review by joenash — APPROVED** [review]
Looks great! Couple of minor copyediting comments

**Comment by nmadadair** [comment]
For the narrativization, once the player has completed this objective they will be given the Divination Spell to open the Lovelace library. Is the reward for this simply that they get the spell rather than the actual task of the objective being that they are creating/generating/uncovering the spell?

Linking previous narrative notes on this objective. How far have we diverged from this original idea?

>  This is the process of actually making the spell. This would be the place where you're not just logging, but doing stuff with the data. This is the place where we'd thinking about how to "test out your spell for the first time." This is an important fictional moment. Then, how does the door open? Divination Spell performs thought bubble comes up with the player saying that they've divined the password? The performing of the spell is really about the testing of the spell. We'd want some sort of narrative popup about the spell to contextualize why divination on the door opens the door. Thought bubble about having requested the data / door's end point / passcode with some code formatting, etc.

**Comment by MarvelProgramming** [comment]
> For the narrativization, once the player has completed this objective they will be given the Divination Spell to open the Lovelace library. Is the reward for this simply that they get the spell rather than the actual task of the objective being that they are creating/generating/uncovering the spell?

I think we could still get away with doing both (i.e. rewarding the player with the spell while still tying in the idea that they're making it as part of the objective)

> > This is the process of actually making the spell. This would be the place where you're not just logging, but doing stuff with the data.

At least when it comes to asking the player to perform more involved work, I think the objective still hits the mark (more on the process of making the spell below)

> > This is the place where we'd thinking about how to "test out your spell for the first time." This is an important fictional moment. Then, how does the door open? Divination Spell performs thought bubble comes up with the player saying that they've divined the password? The performing of the spell is really about the testing of the spell. We'd want some sort of narrative popup about the spell to contextualize why divination on the door opens the door. Thought bubble about having requested the data / door's end point / passcode with some code formatting, etc.

Are some of these parts meant to happen outside of the objective (i.e. not as part of the hacking interface)? When you mention "thought bubbles", I imagine we'd handle that similarly to what the player sees when they initially board the Fog Owl. For example, listening for the `objectiveCompleted` event, locking player movement, displaying some of that "internal dialogue", panning the camera combined with more dialogue, ect.

Perhaps we could make the processing (i.e. filtering the authors) part of the spells creation? For example, filtering the list of authors down uncovers an interesting "coincidence" where by speaking the names of those authors that remain, unlocks a hidden compartment in the statue that has a scroll with the incantation.  We _could_ have the "opening of the library door" also be the "test", since the player will theoretically be using spells to open doors belonging to other houses later on. After the player casts the spell on the door for the first time, we could introduce further dialogue if need be. Let me know what you think of this!

**Comment by nmadadair** [comment]
Responding from my phone quickly (and will give a more thorough reply tomorrow): okay, awesome, I like the idea that we can still have them "build" the spell. I will wait tomorrow for your updates to the description to reflect on and then draft further narrative context that we might include. 

We can definitely add an Operator observation to help with this when the player completes the fourth objective and then "have" the spell.

**Comment by nmadadair** [comment]
If you haven't already merged in the 4th Objective, feel free to do so whenever you have finished it. Then when I start early tomorrow I'll read over and make some additions if needed!

**Pull request merged**